feat: support for custom message generation in `ZddMessageConfigInterface`

B-Galati opened the pull request:
This PR introduces optional custom message generation through a new `CustomMessageGeneratorInterface`,      
  allowing users to have full control over how specific message instances are created.                        
                                                                                                              
  ### Changes                                                                                                  
                                                                                                              
  - Added `CustomMessageGeneratorInterface` with `generateCustomMessage()` method                             
  - Extracted `MessageGenerator` class to centralize message creation logic                                   
  - Updated `ZddMessageFactory` to delegate message generation to `MessageGenerator`                          
  - Added documentation and example usage in README                                                           
                                                                                                              
  ### Notes                                                                                                    
                                                                                                              
  - `CustomMessageGeneratorInterface` is optional and separate from `ZddMessageConfigInterface`               
  - Users can implement both interfaces in the same class if they need custom instantiation                   
  - Custom message instantiation is useful when the default reflection-based approach is insufficient or when 
  messages require specific constructor logic

B-Galati committed:
feat: support for custom message generation in `ZddMessageConfigInterface`

B-Galati committed:
docs: update README wording and fix `message_config_service` registration

ismail1432 commented on src/Factory/ZddMessageFactory.php:
```diff
@@ -26,9 +28,12 @@ public function create(string $className): ZddMessage
-        foreach ($propertyList->getProperties() as $property) {
-            $this->forcePropertyValue($message, $property->name, $property->value);
+        $message = $this->config->generateCustomMessage($className);
+        if (null === $message) {
```
You should precise in phpdoc of `ZddMessageConfigInterface::generateCustomMessage` that the method should return `null` when they want to use the reflection otherwise it will not work.

ismail1432 commented:
Where Am I? 😭 

<img width="916" height="307" alt="Capture d’écran 2025-11-21 à 16 33 54" src="https://github.com/user-attachments/assets/b24d7e2a-a2ab-4bad-813a-378348eea179" />

B-Galati commented on README.md:
````diff
@@ -49,29 +52,44 @@ class MessageConfig implements ZddMessageConfigInterface
 ```yaml
 # config/packages/zdd_message.yaml
   zdd_message:
-    serialized_messages_dir: 'var/serialized_messages' # The directory where the serialized messages will be stored (default: '%kernel.logs_dir%')
````
I added the documentation about it :+1:

ismail1432 reviewed (approved):
LGTM! TY @B-Galati

B-Galati committed:
refactor: extract `MessageGenerator` from `ZddMessageFactory`

This refacto will clearly split responsibilities between these 2 concerns.
Moreover, it will enable the introduction of the custom message feature to handle edge cases.

B-Galati committed:
feat: extract `generateCustomMessage` into optional `CustomMessageGeneratorInterface`

B-Galati committed:
doc: add phpstan target to Makefile help message

B-Galati committed:
chore: remove useless PHPStan ignore

TBoileau commented on src/Factory/Property.php:
```diff
@@ -4,7 +4,7 @@
 final class Property
 {
-    public function __construct(public readonly string $name, public readonly string $type, public readonly mixed $value)
+    public function __construct(public readonly string $name, public readonly string $type, public readonly bool $isNullable)
```
It's a matter of taste, but I prefer `nullable` to `isNullable`. The `is` often seems unnecessary to me when there is already strong typing.

B-Galati replied:
I find this naming more intuitive when doing `if` statement for example: I don't need to inspect the property type to be sure of what I am doing.

TBoileau replied:
That's true 👍  and it's even more so in the context of a library !

TBoileau reviewed (approved):
Thanks for the work !

B-Galati commented:
The new interface is marked as experimental after discussion with @Inkod.

And we also added some warning about nullable cases that can hide ZDD problems for custom messages.
